sonar-scanner-engine/src/main/java/org/sonar/scanner/scan/filesystem/ByteCharsetDetector.java

```diff
@@ -0,0 +1,119 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.scanner.scan.filesystem;
+
+import java.nio.charset.Charset;
+import java.util.Arrays;
+import java.util.LinkedHashSet;
+import java.util.Set;
+
+import static java.nio.charset.StandardCharsets.*;
+
+import javax.annotation.CheckForNull;
+
+import org.apache.commons.io.ByteOrderMark;
+import org.sonar.scanner.scan.filesystem.CharsetValidation.Result;
+import org.sonar.scanner.scan.filesystem.CharsetValidation.Valid;
+
+public class ByteCharsetDetector {
+  // these needs to be sorted by longer first!
+  private static final ByteOrderMark[] boms = {ByteOrderMark.UTF_8, ByteOrderMark.UTF_32LE, ByteOrderMark.UTF_32BE,
+    ByteOrderMark.UTF_16LE, ByteOrderMark.UTF_16BE};
+
+  private Charset userConfiguration;
+  private CharsetValidation validator;
+
+  public ByteCharsetDetector(CharsetValidation validator, Charset userConfiguration) {
+    this.validator = validator;
+    this.userConfiguration = userConfiguration;
+  }
+
+  @CheckForNull
+  public Charset detect(byte[] buf) {
+    Set<Charset> charsets = new LinkedHashSet<>();
+    charsets.add(userConfiguration);
+    charsets.add(Charset.defaultCharset());
+
+    // Try UTF-8 first since we are very confident in it if it's a yes.
+    // Fail if we see nulls to not have FPs if the text is ASCII encoded in UTF-16.
+    Result utf8Result = validator.isUTF8(buf, true);
+    if (utf8Result.valid() == Valid.YES) {
+      return utf8Result.charset();
+    } else if (utf8Result.valid() == Valid.MAYBE) {
+      return detectAscii(buf, charsets);
+    }
+
+    // try UTF16 with both endiness. Fail if we see nulls to not have FPs if it's UTF-32.
+    Result utf16 = validator.isUTF16(buf, true);
+    if (utf16.valid() == Valid.YES) {
+      if (validator.isValidUTF16(buf, UTF_16LE.equals(utf16.charset()))) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Merge this if statement with the enclosing one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://next.sonarqube.com/sonarqube/coding_rules#rule_key=squid%3AS1066)